### PR TITLE
gh-118124: Use static_assert() in Py_BUILD_ASSERT() on C11

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-118124

### CODE REVIEW

**Comment by vstinner** [comment]
See also PR gh-118125.

File: Include/pymacro.h [hunk 1]
```diff
@@ -46,24 +46,37 @@
...
-
-   #define foo_to_char(foo)  \
-       ((char *)(foo)        \
-        + Py_BUILD_ASSERT_EXPR(offsetof(struct foo, string) == 0))
-
-   Written by Rusty Russell, public domain, http://ccodearchive.net/ */
-#define Py_BUILD_ASSERT_EXPR(cond) \
+#if __STDC_VERSION__ >= 201112L
+#  define Py_BUILD_ASSERT_EXPR(cond) \
+    (sizeof(struct { _Static_assert(cond, #cond); int dummy; }), \
```
Comment (namniav) [inline Include/pymacro.h]: Defining structs in expression is not valid C++. And `_Static_assert` is not a keyword in C++.

**Review by namniav — CHANGES_REQUESTED** [review]

File: Include/pymacro.h (hunk 1 above)
Comment (vstinner) [reply Include/pymacro.h]: Is `__STDC_VERSION__` defined in C++?

**Comment by vstinner** [comment]
Code based on https://stackoverflow.com/questions/51388341/is-there-any-way-to-slip-a-static-assert-into-an-expression-in-iso-c11 answer.

File: Include/pymacro.h (hunk 1 above)
Comment (vstinner) [reply Include/pymacro.h]: For C++, we already have this code:

```c
// gh-91782: On FreeBSD 12, if the _POSIX_C_SOURCE and _XOPEN_SOURCE macros are
// defined, <sys/cdefs.h> disables C11 support and <assert.h> does not define
// the static_assert() macro.
// https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=255290
//
// macOS <= 10.10 doesn't define static_assert in assert.h at all despite
// having C11 compiler support.
//
// static_assert is defined in glibc from version 2.16. Compiler support for
// the C11 _Static_assert keyword is in gcc >= 4.6.
//
// MSVC makes static_assert a keyword in C11-17, contrary to the standards.
//
// In C++11 and C2x, static_assert is a keyword, redefining is undefined
// behaviour. So only define if building as C (if __STDC_VERSION__ is defined),
// not C++, and only for C11-17.
#if !defined(static_assert) && (defined(__GNUC__) || defined(__clang__)) \
     && defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L \
     && __STDC_VERSION__ <= 201710L
#  define static_assert _Static_assert
#endif
```

Comment (vstinner) [reply Include/pymacro.h]: I updated the PR to test explicitly for `defined(__STDC_VERSION__)`.

Comment (namniav) [reply Include/pymacro.h]: > Is `__STDC_VERSION__` defined in C++?

Ah, no. Missed that.

**Comment by namniav** [comment]
Do you prefer leaving the macros unchanged for all C++ and C99, for complexity?

Since you're working on this, can/should I close my PR?

**Comment by vstinner** [comment]
> Do you prefer leaving the macros unchanged for all C++ and C99, for reducing complexity?

_Static_assert() was added to C11. I don't think that we can expect it to be available on C99.

Is it available on C++?

**Comment by namniav** [comment]
> _Static_assert() was added to C11. I don't think that we can expect it to be available on C99.
>
>  Is it available on C++?

No. `static_assert` was added to C++11. 

VLA was added to C99 (optional since C11) and is often available as compiler extension for older C versions and most C++ versions. There're tricks to also fix the two macros for them but would introduce some complexity in code, at least not as clean as your PR.

**Comment by vstinner** [comment]
> No. static_assert was added to C++11.

I updated the PR for C++. It now uses static_assert() on C++11 and newer.

**Comment by vstinner** [comment]
The Docs CI job failed with:

```
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Doc/tools/check-warnings.py", line 311, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/runner/work/cpython/cpython/Doc/tools/check-warnings.py", line 279, in main
    with Path("Doc/sphinx-warnings.txt").open(encoding="UTF-8") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.3/x64/lib/python3.12/pathlib.py", line 1013, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'Doc/sphinx-warnings.txt'
```

**Comment by vstinner** [comment]
> The Docs CI job failed

I rebased my PR on the main branch to retrieve a fix. I also squashed my commits.

**Comment by vstinner** [comment]
@namniav: So what do you think of this approach?

**Comment by namniav** [comment]
It looks good to me. I'm not sure whether we should fix Py_BUILD_ASSERT_EXPR for older C versions and C++, I'm ok with either though.

**Review by namniav — APPROVED** [review]

**Comment by vstinner** [comment]
> It looks good to me. I'm not sure whether we should fix Py_BUILD_ASSERT_EXPR for older C versions and C++, I'm ok with either though.

An alternative is to remove it, but it seems like it's used outside by 3rd party projects.

**Pull request merged**